### ISSUE TITLE
[FEATURE] Zero-Inflated Regressor

### ISSUE THREAD

**Issue by Garve** [koaning/scikit-lego#441]
Hi again!

I propose a meta estimator for dealing with zero-inflated datasets, i.e. where the labels y are zero quite often. The use-case is there since there is a large literature about zero-inflated Poisson models for counting, for example, as in [the paper "Predicting property damage from tornadoes with zero-inflated neuralnetworks"](https://reader.elsevier.com/reader/sd/pii/S2212094718301385?token=BDD2ECB6F61368C81CD333D20D9A1E8A0A1112ECA7D9F9AE4520460F714E7CC56845725E144F2971E7C31A01A6CFA58B).

[I programmed a prototype for my toy library again](https://github.com/Garve/scikit-bonus/blob/master/skbonus/meta/_zero_inflated_regressor.py), but I think it might be interesting for more people, too. So far, it works like this:

```python
import numpy as np
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.model_selection import cross_val_score
from skbonus.meta import ZeroInflatedRegressor

np.random.seed(0)
X = np.random.randn(10000, 4)
y = ((X[:, 0]>0) & (X[:, 1]>0)) * np.abs(X[:, 2] * X[:, 3]**2) # many zeroes here, in about 75% of the cases.

zir = ZeroInflatedRegressor(
    classifier=RandomForestClassifier(random_state=0),
    regressor=RandomForestRegressor(random_state=0)
)

print('ZIR(RFC+RFR) r²:', cross_val_score(zir, X, y).mean())
print('RFR r²:', cross_val_score(RandomForestRegressor(random_state=0), X, y).mean())
```

Output:
```bash
ZIR(RFC+RFR) r²: 0.8992399247461955
RFR r²: 0.8516520051101274
```

Under the hood, a classifier is trained to see if the output should be zero. After this, a regressor is training on only the samples with a non-zero label.
Such a meta regressor is especially useful for simpler regressors that have problems with zero-inflated datasets, such as plain linear regression.

Best
Robert

**Comment by koaning** [comment]
This sounds right up our alley in my mind. @MBrouns any comments?

**Comment by MBrouns** [comment]
Love it! Tiny nitpick on your implementation: currently you apply the regressor to all samples and multiply by zero in case the classifier thinks it's a zero. I think it'd be nicer if the zero predictions were filtered from the input of the regressor to save a bit of computation time

**Comment by Garve** [comment]
Happy to hear that :)

And true, I will fix it later!

Best
Robert
